### ISSUE TITLE
Installation problem

### ISSUE THREAD

**Issue by miguelem** [fblascogarma/google_ads_backend#1]
When I'm doing the migration step:

`python manage.py migrate`

It doesn't do any migrations. It shows a long list of messages. 

The first one is:

`Traceback (most recent call last):`

And the last ones are:

```
File "C:\Users\usuario\AppData\Local\Programs\Python\Python310\lib\site-packages\google\rpc\error_details_pb2.py", line 39, in <module>
    _RETRYINFO = DESCRIPTOR.message_types_by_name["RetryInfo"]
AttributeError: 'NoneType' object has no attribute 'message_types_by_name'
```

The next step (create superuser) does a similar thing and doesn't create the superuser. 

The last step (runserver) works and it loads the page OK, but the console says: "You have 70 unapplied migration(s). "

I'm a newbie, so it's probably a problem with my config. But I've been able to install Django using other tutorial and apply migrations, so I'm confused here. Any tips or links would be appreciated. 

Thanks!

**Comment by fblascogarma** [comment]
Hi @miguelem , thank you for your question!

NoneType means that instead of an instance of whatever Class or Object you think you're working with, you've actually got None. That usually means that an assignment or function call up above failed or returned an unexpected result.

Can you paste the entire response message when you try to make the migration, please?

Also, [here](https://youtu.be/WvP3vtc_o7Q) is a YouTube video tutorial that shows every step that you can use to compare what you are getting versus what I got in the video.

**Comment by miguelem** [comment]
Thanks @fblascogarma, but I decided to try a different solution (I'm not familiar with React, so I will try to use just Python/Django and HTML/CSS, no frontend frameworks). 

Do you think I could reuse some things from your repositories & docs, or it will be easier to just use the official Google Ads client library for Python?  

(I'm a Python/Django beginner and I want the API mainly for reporting)

**Comment by fblascogarma** [comment]
You can most definitely reuse most of the code. The backend repo has all the functions that make API calls to Google Ads, and it's in Python. 

If you want to use the API mainly for reporting, you will find very helpful the [get_campaigns.py](https://github.com/fblascogarma/google_ads_backend/blob/main/api/get_campaigns.py) file that has the function to get campaign performance metrics for all campaign types (Search, Shopping, Display, Video, etc.) of a Google Ads account. Furthermore, if you have Smart Campaigns in the account and you want to see the search terms report, the [get_search_terms_report.py](https://github.com/fblascogarma/google_ads_backend/blob/main/api/get_search_terms_report.py) file has the function to get that data.

Also, if you want data of an account, you will find very helpful the [list_accessible_accounts.py](https://github.com/fblascogarma/google_ads_backend/blob/main/api/list_accessible_accounts.py) file. Lastly, if you want to see the billing information of a Google Ads account, the [get_billing_info.py](https://github.com/fblascogarma/google_ads_backend/blob/main/api/get_billing_info.py) file has the function to get relevant data around that.

Let me know if this was helpful or if you have any further questions.

I'm closing this issue as it's not an issue anymore.

Happy hacking, @miguelem !

**Comment by fblascogarma** [comment]
Hi @miguelem , can you try the installation steps again but this time do 

python manage.py makemigrations

And then do

python manage.py migrate

and let me know, please?

**Comment by miguelem** [comment]
Thanks @fblascogarma!

I've tried the installation steps again from scratch, following your video (only for the backend, using Windows), and when I try the makemigrations or migrate lines and it gives me:

`AttributeError: 'NoneType' object has no attribute 'message_types_by_name'`

---------------------

Also, I want to make you another question: I've managed to get the list of campaigns using a modified version of the official client library example (https://github.com/googleads/google-ads-python/blob/main/examples/basic_operations/get_campaigns.py).

I get something like this in my browser:

```
customer {
  resource_name: "customers/XXX"
  id: XXX
}
campaign {
  resource_name: "customers/XXX/campaigns/XXX"
  status: ENABLED
  name: "XXX"
  id: XXX
}
metrics {
  impressions: 0
}
```

But that's not really what I need. I want to "process" this data with Python/Django (take/store the data from those fields, make some calculations...) and then show the "processed" data to the user on an HTML table. 

After having a quick look at your code, it seems like most of the data is not stored on Django. For example, for many fields of the get_campaigns.py or get_search_terms_report.py queries, I don't see anything related on the models.py, no "save()" lines anywhere...

I don't know if it's because:
A) You store the data in a database in a way that I don't understand (maybe React-related?) 
B) You don't store most of the data on a database, you just store it temporarily in Python variables and then show it to the user, with no "long-term storage" for that data. 

I'm a beginner, so I'm probably missing something really obvious. I want to understand not just how your code works, but also what are the best practices for working with Google Ads API data. 

For example, I want to use the API mostly for reporting (for different clients/accounts), so my first instinct is to store all their report data on the database (maybe easier to process that data, faster to load the reports, fewer API requests...). But I don't know, maybe not using the database and calling the data always directly from the API (+ some Python code to process it) has some clear advantages (easier to configure than "multi-tenant" databases with many different reports from many different clients, more secure...). 

It would be great if you could give me your input or send me some links where I can read or ask about this. 

Thanks again!

**Comment by fblascogarma** [comment]
Hi @miguelem , thank you for your question. I don't save the campaign performance data anywhere. I just send it to the frontend to be displayed as a table for users to see. It's a design decision I made to save memory. Furthermore, users generally use date filters to see the campaign performance data, so in my opinion it's better to have a dynamic approach instead of a static one. 

Check out the [Reporting section](https://docs.google.com/document/d/10WyzDUNZTVHhenWgCopf3YyaRDsyoavc7aHuOAMJF6Y/edit#bookmark=kix.ox5j48btuadj) of the technical guide I wrote to see how I show the data on the frontend.

Let me know if you have additional questions.
